Quickfix integration buggy for multi-line errors

Seems like for multi-line error messages the plugin will populate the quickfix buffer with the message as-is, which means for an N-line long error message (Neo)Vim thinks there are N errors. For example if the error message is (snipped):

```haskell
Main.hs:18:5: error:
    • Couldn't match type ‘b1’ with ‘Zipper g b'0 a'0’
      ‘b1’ is a rigid type variable bound by
        the type signature for:
          fmap :: forall a b1. (a -> b1) -> Zipper g b a -> Zipper g b b1
        at Main.hs:16:3
```

The quickfix buffer thinks the following are errors (numbered):

1. || [1 of 1] Compiling Main ( Main.hs, interpreted ) 
2. Main.hs|18 col 5 error| error: • Couldn't match type ‘b1’ with ‘Zipper g b'0 a'0’
3. || ‘b1’ is a rigid type variable bound by 
4. || the type signature for:
5. || fmap :: forall a b1. (a -> b1) -> Zipper g b a -> Zipper g b b1 
6. || at Main.hs:16:3

--- a/plugin/ghci.vim
+++ b/plugin/ghci.vim
@@ -32,18 +32,25 @@ command! -nargs=0 -bang GhciRestart call ghci#process#restart()
 
 if g:ghci_use_neomake
     " Neomake integration
+
+    " Try GHC 8 errors and warnings, then GHC 7 errors and warnings, and regard
+    " lines starting with two spaces as continuations on an error message. All
+    " other lines are disregarded. This gives a clean one-line-per-entry in the
+    " QuickFix list.
+    let s:efm = '%E%f:%l:%c:\ error:%#,' .
+                \ '%W%f:%l:%c:\ warning:%#,' .
+                \ '%f:%l:%c:\ %trror: %m,' .
+                \ '%f:%l:%c:\ %tarning: %m,' .
+                \ '%E%f:%l:%c:%#,' .
+                \ '%E%f:%l:%c:%m,' .
+                \ '%W%f:%l:%c:\ Warning:%#,' .
+                \ '%C\ \ %m%#,' .
+                \ '%-G%.%#'
+
     let g:neomake_ghci_maker = {
             \ 'exe': 'cat',
             \ 'args': [ghci#maker#get_log_file()],
-            \ 'errorformat':
-                \ '%-G%\s%#,' .
-                \ '%f:%l:%c:%trror: %m,' .
-                \ '%f:%l:%c:%tarning: %m,'.
-                \ '%f:%l:%c: %trror: %m,' .
-                \ '%f:%l:%c: %tarning: %m,' .
-                \ '%E%f:%l:%c:%m,' .
-                \ '%E%f:%l:%c:,' .
-                \ '%Z%m'
+            \ 'errorformat': s:efm
         \ }
 endif
 

--- a/plugin/ghci.vim
+++ b/plugin/ghci.vim
@@ -39,6 +39,7 @@ if g:ghci_use_neomake
     " QuickFix list.
     let s:efm = '%E%f:%l:%c:\ error:%#,' .
                 \ '%W%f:%l:%c:\ warning:%#,' .
+                \ '%W%f:%l:%c:\ warning:\ [-W%.%#]%#,' .
                 \ '%f:%l:%c:\ %trror: %m,' .
                 \ '%f:%l:%c:\ %tarning: %m,' .
                 \ '%E%f:%l:%c:%#,' .

